### PR TITLE
fix promotion by avoiding integer division in mse and crossentropy

### LINKED ISSUE #514: Gradient of `mse` is of type Float64 for data of type Float32
```
args = Flux.param.((CuArray(rand(Float32,32, 32, 1, 10)), CuArray(rand(Float32, 32, 32, 1, 10))))
y, back = Flux.Tracker.forward(Flux.mse, args...)
julia> typeof(y)
Flux.Tracker.TrackedReal{Float32}

julia> typeof(back(1.0f0))
Tuple{TrackedArray{…,CuArray{Float64,4}},TrackedArray{…,CuArray{Float64,4}}}
```

This seems to be due to the division by an integer in `mse`, since `inv(::Int)::Float64`

This worked for me as a workaround:
```
mse(ŷ, y) = sum((ŷ .- y).^2) / convert(Float32, length(y))
```

but hopefully we can fix this generally.

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes #514

### CODE REVIEW

**Comment by MikeInnes** [comment]
Nice trick. Do you know of any downsides to doing things this way?

I think testing this is particularly important given how tempting it's going to be to simplify it.

**Comment by KristofferC** [comment]
> Do you know of any downsides to doing things this way?

Not really. I've run with this PR a bit both on CPU and GPU and it seems to work ok. 

> I think testing this is particularly important given how tempting it's going to be to simplify it.

Agreed, will add some.

**Comment by KristofferC** [comment]
Added tests.

**Comment by KristofferC** [comment]
Wait, this test doesn't test the correct thing!

**Comment by KristofferC** [comment]
It seems there is still spurious promotion when using onehot vectors as input (see the `@test_broken` in last commit.

**Comment by KristofferC** [comment]
That promotion seems to also exist in the forward pass:

```
julia> g(ŷ, y) = sum(ŷ - y)
g (generic function with 1 method)

julia> typeof.((ŷ, y))
(Flux.OneHotVector, Array{Float32,1})

julia> fwd, back = Flux.Tracker.forward(g, ŷ, y);

julia> typeof(fwd)
Flux.Tracker.TrackedReal{Float64}
```

**Comment by KristofferC** [comment]
I added a fix to the onehot promotion issue by mapping the onehot arrays to Bools instead of Float64.

File: test/layers/stateless.jl [hunk 1]
```diff
@@ -49,4 +49,17 @@ const ϵ = 1e-7
   @testset "logitbinarycrossentropy" begin
     @test logitbinarycrossentropy.(logŷ, y) ≈ binarycrossentropy.(σ.(logŷ), y; ϵ=0)
   end
+
+  @testset "no spurious promotions" begin
+    for T in (Float16, Float32, Float64)
+      y = rand(T, 2)
+      for ŷ in (rand(T, 2), Flux.onehot(2, [1, 2]))
+          for f in (mse, crossentropy, logitcrossentropy)
+              fwd, back = Flux.Tracker.forward(mse, ŷ, y)
```
Comment (KristofferC) [inline test/layers/stateless.jl]: `mse -> f`

**Comment by KristofferC** [comment]
Actually, mapping to `Bool` doesn't work in cases where we do a `log` on the onehot vector (for example in `crossentropy`. The `Bool` will be converted to a `Float64`.

**Comment by staticfloat** [comment]
@MikeInnes is there anything else you're interested in about this?  If not, I'd like to merge this in.

**Comment by MikeInnes** [comment]
The `param` change isn't necessary to fix `crossentropy`, which works fine with a tracked array and a one hot vector as input. The fix is also not ideal since it won't work with in-place gradients. Given that one hot vectors are generally not turned into parameters anyway I suggest just removing that code and we'll merge the basic fix in.

[For some context, the issue with `TrackedArray{<:Integer}` is that gradients may not be integers, which causes problems; hence the conversion to floating point both for one hot vectors and things like `param([1, 2, 3])`.)

**Comment by KristofferC** [comment]
I think I did the changes requested.


> For some context, the issue with TrackedArray{<:Integer} is that gradients may not be integers, which causes problems

We should probably add a test for this problem then since this PR passed even with the new param def.

**Comment by MikeInnes** [comment]
:+1: Thanks!

**Pull request merged**